include/pybind11/stl.h

```diff
@@ -144,7 +144,7 @@ template <typename Type, typename Value> struct list_caster {
     using value_conv = make_caster<Value>;
 
     bool load(handle src, bool convert) {
-        if (!isinstance<sequence>(src) || isinstance<str>(src))
+        if (!isinstance<sequence>(src) || isinstance<str>(src) || isinstance<native_str>(src))
```

Comment:
redundant check on Python 3.x?

Reply:
Yes, exactly.
I did it this way mostly because I thought it's the clearest way of putting the condition, for the purpose of this code review. ("optimize later")

To also answer @bstaletic question here:
> I still don't understand. Why does it make sense to allow `bytes` on python3, but not on python2? I would understand, were pybind11 dealing in `py::native_bytes` and `py::native_str` everywhere. It isn't, so having `py::native_str` in this one place seems very surprising.

TL;DR: What would be the behavior that you prefer to see here?

With Python 2 the interpreter (!) will tell you `bytes is str` is `True`, therefore you cannot enable conversion of `bytes` to `list` without also enabling conversion of [Python 2 native `str`] to `list`. That's a feature you can only have with Python 3. — Another way to put it: do you NOT want to use the feature in Python 3 because you cannot have it in Python 2?

We can do anything else we want here, what you see is what I assumed was the original intent.
(At some point long time ago, when there was only Python 2, I had a similar problem and disabled `str` to `list` conversions to prevent confusing accidents.)
Once we are certain what we want, we could optimize, e.g. by putting the #ifdef here and not even having the `native_str` type.

Reply:
> TL;DR: What would be the behavior that you prefer to see here?

Either:

- Ban everything that's string-like in the `list_caster`
- Or allow `bytes` in both py2 and py3

> With Python 2 the interpreter (!) will tell you `bytes is str` is `True`, therefore you cannot enable conversion of `bytes` to `list` without also enabling conversion of [Python 2 native `str`] to `list`. 

Why is py2 `str` (or py2 `bytes`, but those are the exact same thing) to `list` problematic if py3 `bytes` to `list` isn't?

> Another way to put it: do you NOT want to use the feature in Python 3 because you cannot have it in Python 2?

Rarely, but if it helps me reason about what is going on in the code, I see nothing wrong with that.

Reply:
> Ban everything that's string-like in the list_caster

That's current behavior. Two consequences:
1. Effectively you do not want to use a feature that's feasible with Python 3 because it's infeasible in Python 2.
2. Effectively a rejection of #2198. 

> Or allow bytes in both py2 and py3

That breaks the existing test in test_stl.py: `assert m.func_with_string_or_vector_string_arg_overload('A') == 3`
We'd have to change that to u'A', `list('A')`, or `iter('A')` to pass again.

The behavior currently implemented in this PR means:
* If you care about Python 3 (and the future) only: you're fine!
* If you still care about Python 2 even past EOL, you'll have to use explicit `list(s)` or `iter(s)` (which is also Python 3 compatible).

I feel that's the best compromise, but I'll go with the votes of the team.
@YannickJadoul and Wenzel, what are your votes on this specific `list_caster` question?

Reply:
> That breaks the existing test in test_stl.py: assert `m.func_with_string_or_vector_string_arg_overload('A') == 3`

Is this an actual concern, considering that this pull request already contains a breaking change.

Reply:
> > That breaks the existing test in test_stl.py: assert `m.func_with_string_or_vector_string_arg_overload('A') == 3`
> 
> Is this an actual concern, considering that this pull request already contains a breaking change.

I'd say so, I think it's worth getting each of the 3 questions settled in the best way possible.

The only fully intentional change is for `isinstance<str>`. We could preserve the behaviors for the other 2, although I've not looked into what it would take to preserve the implicit `bytes` to `actual unicode` conversions. I'm waiting for Yannick's and Wenzel's takes before investing time there.